Rename powershell.exe to pwsh.exe

SteveL-MSFT opened the pull request:
renamed powershell.exe to pwsh.exe
updated powershell core code dependencies, console host references, and tests
updated vscode json files
updated docker files
updated logic in moduleintrinsics.cs to account for SxS where you are starting pwsh from powershell.exe

Fix https://github.com/PowerShell/PowerShell/issues/4214
<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to [CONTRIBUTING.md](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md).

-->

SteveL-MSFT committed:
[feature]
renamed powershell.exe to pwsh.exe

daxian-dbw commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -418,7 +418,7 @@ Describe "TabCompletion" -Tags CI {
                 @{ inputStr = 'gcm -Module PackageManagement Find-Pac'; expected = 'Find-Package'; setup = $null }
                 @{ inputStr = 'ipmo PackageM'; expected = 'PackageManagement'; setup = $null }
-                @{ inputStr = 'Get-Process powersh'; expected = 'powershell'; setup = $null }
+                @{ inputStr = 'Get-Process pws'; expected = 'pwsh'; setup = $null }
```
A typo -- `pws`

SteveL-MSFT replied:
This is for the tab completion test so `pws` completes to `pwsh` since there isn't necessarily a `powershell` process anymore and `pw` might resolve to something else

daxian-dbw replied:
Ah, yes, my mistake.

SteveL-MSFT commented on demos/SSHRemoting/README.md:
```diff
@@ -173,7 +173,7 @@ Microsoft Windows [Version 10.0.10586]
 # Windows to Windows
 #
-C:\Users\PSUser\Documents>"C:\Program Files\PowerShell\6.0.0.17\powershell.exe"
+C:\Users\PSUser\Documents>pwsh.exe
```
We should update the msi installer to add it to %PATH%, I wanted to remove it since it references a specific version path.  I'll do the msi update separate from this PR.

SteveL-MSFT commented:
rebased to fix merge conflict

SteveL-MSFT committed:
[feature]
fixed check in appveyor.psm1

PaulHigin commented on src/Microsoft.PowerShell.ScheduledJob/ScheduledJobDefinition.cs:
```diff
@@ -42,7 +42,7 @@ public sealed class ScheduledJobDefinition : ISerializable, IDisposable
 
         // Task Action strings.
-        private const string TaskExecutionPath = @"powershell.exe";
+        private const string TaskExecutionPath = @"pwsh.exe";
```
This code assumes a discoverable pwsh.exe native command.  Does this fix include adding pwsh.exe to the system path environment variable?  If so what version?  I wonder if we should allow the user to select a powershell.exe or pwsh.exe path to use for scheduled jobs (maybe in a separate issue/pr).

PaulHigin replied:
Maybe we should use Utils.DefaultPowerShellAppBase here as we do with PowerShellProcessInstance.cs to ensure the scheduled job is using the correct path.
The problem with scheduled jobs is that they can run indefinitely on a machine and if pwsh.exe is uninstalled or installed to a new location then they will break.  Of course we have a similar problem with credentials expiring.

SteveL-MSFT replied:
I have a separate PR coming for the msi installer to append to %PATH% which would also solve the Win+R->pwsh usage

If PSCore6 is uninstalled, the full path won't help either.

SteveL-MSFT replied:
Actually, I'll do the %PATH% msi change as part of this PR as it's more useful that pwsh.exe differs form powershell.exe and I can put it at the end of the path rather than taking over the front

PaulHigin commented on src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs:
```diff
@@ -3166,7 +3166,7 @@ private void CreateContainerProcessInternal()
                     //
                     cmd = string.Format(System.Globalization.CultureInfo.InvariantCulture,
-                        @"{{""CommandLine"": ""powershell.exe {0} -NoLogo {1}"",""RestrictedToken"": {2}}}",
+                        @"{{""CommandLine"": ""pwsh.exe {0} -NoLogo {1}"",""RestrictedToken"": {2}}}",
```
Same issue here in that we rely on a "built-in" pwsh.exe command that is discoverable.  But we have the same issue with ssh.exe used for SSH based remoting.

PaulHigin replied:
Actually as I think about it, we don't want to change this here.  This is a command run on a target VM and relies on the built-in Windows PowerShell.  pwsh.exe will not be available on a new VM.

SteveL-MSFT replied:
So this command is run within the VM expecting Windows PowerShell?

PaulHigin replied:
Yes

SteveL-MSFT replied:
Ok, will change back.

SteveL-MSFT committed:
[feature]
updated msi to include pwsh in path and app paths
reverted change for hyper-v powershell direct
reapplied test changes reverted by rebase

PaulHigin reviewed (approved):
LGTM

SteveL-MSFT commented:
@daxian-dbw can you review again since last commit?

daxian-dbw commented on src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:
```diff
@@ -791,8 +793,9 @@ private static string RemoveSxSPsHomeModulePath(string currentProcessModulePath,
                     string parentDir = Path.GetDirectoryName(trimedPath);
                     string psExePath = Path.Combine(parentDir, powershellExeName);
+                    string oldExePath = Path.Combine(parentDir, oldPowershellExeName);
                     string psDepsPath = Path.Combine(parentDir, "powershell.deps.json");
```
Once you change it to `pwsh` in csproj, I guess the .deps.json file will be changed to `pwsh.deps.json`.
But maybe you want to keep the check for `powershell.deps.json` for now just like the check for old "powershell.exe".

SteveL-MSFT replied:
Yeah, it did get changed to `pwsh.deps.json`, will add `oldDepsPath`

SteveL-MSFT replied:
Do we really need the deps.json check?

daxian-dbw replied:
I remember we didn't have the .deps.json check initially and it caused a build break and the fix was to add this deps.json check. Here is the related issue #1744 and #1792

Not sure if that requirement is applicable now.

SteveL-MSFT replied:
I'll keep it for now

daxian-dbw commented on assets/Product.wxs:
```diff
@@ -103,13 +105,23 @@
+            </Component>
+            <!-- add ourselvs to %PATH% so pwsh.exe can be started from Windows PowerShell or cmd.exe -->
+            <Component Id="SetPath" Guid="{9dbb7763-7baf-48e7-b025-3bdedcb0632f}">
+              <Environment Id="PATH" Action="set" Name="PATH" Part="last" Permanent="yes" System="yes" Value="[$(var.ProductVersionWithName)]"/>
```
Do we remove the path when uninstalling the package?

SteveL-MSFT replied:
Will fix

daxian-dbw commented on src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:
```diff
@@ -769,8 +769,10 @@ private static string RemoveSxSPsHomeModulePath(string currentProcessModulePath,
+            const string oldPowershellExeName = "powershell";
 #else
             const string powershellExeName = "pwsh.exe";
+            const string oldPowershellExeName = "powershell.exe";
```
Shall we remove the old exe name when we reach RC/RTM?

SteveL-MSFT replied:
I think this is still needed if we start pwsh from Windows PowerShell

daxian-dbw replied:
I think the "start pwsh from windows powershell" case is handled in [`NeedToClearProcessModulePath`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs#L725), and `RemoveSxSPsHomeModulePath` only handles powershell core.

SteveL-MSFT replied:
Because of the .deps.json file check, this doesn't apply to Windows PowerShell anyways.  We can remove at RC/GA

daxian-dbw replied:
Then I will open an issue on this.

daxian-dbw commented on src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs:
```diff
@@ -3166,7 +3166,7 @@ private void CreateContainerProcessInternal()
                     //
                     cmd = string.Format(System.Globalization.CultureInfo.InvariantCulture,
-                        @"{{""CommandLine"": ""pwsh.exe {0} -NoLogo {1}"",""RestrictedToken"": {2}}}",
+                        @"{{""CommandLine"": ""powershell.exe {0} -NoLogo {1}"",""RestrictedToken"": {2}}}",
```
A comment explaining why `powershell.exe` is used here will be helpful when looking at this code in future.

daxian-dbw commented:
@SteveL-MSFT One more issue: when creating Unix packages, symbolic link is created to point to `powershell` in `$pshome` (see the following places). You need to update the symbolic link file name and the target file name.

Besides, are you planning to change the package name? (I assume not).

https://github.com/PowerShell/PowerShell/blob/master/tools/packaging/packaging.psm1#L459
https://github.com/PowerShell/PowerShell/blob/master/tools/packaging/packaging.psm1#L471
https://github.com/PowerShell/PowerShell/blob/master/tools/packaging/packaging.psm1#L587

SteveL-MSFT commented:
@daxian-dbw will update the symlink, the package name is not changing

SteveL-MSFT committed:
[feature]
address PR feedback

SteveL-MSFT committed:
[feature]
missed test generating fake .deps.json file
no longer need code in packaging.psm1 to ensure binaries are named correctly (since dotnet does it correctly now)
updated symlinks in packaging, didn't update $name value since it appears to be used for the package name which isn't changing

daxian-dbw committed:
Some more updates to packaging.psm1

daxian-dbw reviewed (changes requested):
I did some more updates to the packaging.psm1 regarding the name change.

I think we also need to update the docker files, including those in `docker/release` and `tools\releaseBuild\Images`. But I'm OK if you want to do that in a separate PR.

SteveL-MSFT commented on src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:
```diff
@@ -791,8 +796,10 @@ private static string RemoveSxSPsHomeModulePath(string currentProcessModulePath,
+                    string oldExePath = Path.Combine(parentDir, oldPowershellExeName);
+                    string psDepsPath = Path.Combine(parentDir, powershellDepsName);
+                    string oldDepsPath = Path.Combine(parentDir, oldPowershellDepsName);
+                    if ((File.Exists(psExePath) && File.Exists(psDepsPath)) || (File.Exists(oldDepsPath) && File.Exists(oldDepsPath)))
```
Will fix

SteveL-MSFT commented:
Let's do the docker updates as separate PR

SteveL-MSFT committed:
[feature]
fix check for SxS

tmknight commented:
Shouldn't this be an alias vs a complete rename?  Please point me to the discussion where the decision to rename took place.  Thanks

markekraus commented:
@tmknight here is where the PowerShell comitee's decision was announced: https://github.com/PowerShell/PowerShell/issues/4214#issuecomment-335989245
